### DISCUSSION TITLE
Revisit using Cloud Exchange Format

### DISCUSSION THREAD

**Discussion by kierank** [vsf-tv/gccg-api#16] [General]
Can we revisit the idea of using the Cloud Exchange Format (8-bit data with 2-bit LSBs at the end of the line). This is a more software-friendly 10-bit approach, and I am not sure it's a great idea to be burning CPU and the planet unpacking pgroups for the next few decades.

The BBC have kindly provided optimised functions for the the Cloud Exchange Format already.

https://github.com/vsf-tv/cef

cc @peterbrightwell

**Comment by peterbrightwell** [comment]
Agreed, broadcasters are concerned about the complexity of using a on-premise/edge interconnect for “second wave” use cases where we connect all-software media functions within a cluster of VMs or containers.  Complexity and power could be reduced with a suitably efficient packing format.

The CEF proposal is just one way of doing things (and we did simplify it somewhat from earlier proposals), but provides a more “software-friendly” alternative to pgroups.

In any case I think we need to have a message to those who are asking this sort of question, such as how does the work happening now help with minimising the friction between dynamically provisioned applications/functions.  That could mean avoiding unnecessary mezzanine compression but also unnecessary unpacking/repacking.

**Comment by jmailhot** [comment]
It is important for a base format to exist that everything can speak.  that said, the document ought to accommodate the case where a sender offers and receiver accepts other formats.  NMOS provides a mechanism for receivers to indicate support for multiple payload types, but its less clear to me the mechanics of how a sender can indicate multiple possibilities or how a controller would tell a sender to use one -vs- another.  Perhaps if we consider the pixel-arrangement scheme a transport parameter rather than a payload type it would be simpler.

**Comment by jmailhot** [comment]
Next logical step would be to write a companion spec (short) that describes CEF (pointing to github for the detail definition) and then describes how to signal and negotiate its use.
* receivers can advertise multiple capabilities (BCP-004)
* senders (today) say what they are sending (when active)
* when senders are not active, is there a way for a sender to indicate what it *could* do and then specify that when the sender is activated?

**Comment by jmailhot** [comment]
looks like the idea is that one would use AMWA IS-11 on the sender side to advertise the supported contraint sets, and the related work on the receiver side to advertise the supported constraints, and then the controlling entity can choose the format and instruct the sender to change its active format before connecting to the receiver.

so we need a well-defined set of AMWA terminology to specify the CEF format?

**Reply by jmailhot** [reply]
next step on this is for @pac-work and andy R to discuss and determine the suitability and gaps around the IS-11 capabilities negotiation approach, and whether that can enable sender to support multiple formats in an offer/ask methodology for this use case.
also need to examine what API changes or additions are needed for the underlying implementation to signal these capabilities up to the application and onward into the advertisements, and the reverse path for puts to the /active branch on the sender.

**Comment by Andy-Rayner** [comment]
So it seems like there are the options of pgroups, CEF, unpacked 8bit or unpacked 16bit
The last 2 seem to be gaining traction in implimentations as the cost of bandwidth is becoming less of an issue and the cost of processing more so.
Has anyone done an IS-11 implimentation that we could talk to/liaise with?

**Reply by rmsporter** [reply]
> Has anyone done an IS-11 implimentation that we could talk to/liaise with?

An IS-11 implementation from Macnica has been contributed to nmos-cpp and is currently under review, in parallel with an NMOS test suite for IS-11. These are currently pull requests in the relevant GitHub repos and should both be merged in the not-too-distant future.

**Comment by peterbrightwell** [comment]
Taking the “bandwidth is becoming less of an issue” argument further still I’ve also heard single-precision floats proposed… This would certainly be future-proof in terms of dynamic range and accuracy, assuming that the floating point processing performance isn’t an issue.

P

From: Andy-Rayner ***@***.***>
Date: Thursday, 23 May 2024 at 15:36
To: vsf-tv/gccg-api ***@***.***>
Cc: Peter Brightwell ***@***.***>, Mention ***@***.***>
Subject: Re: [vsf-tv/gccg-api] Revisit using Cloud Exchange Format (Discussion #16)

External: Think before clicking

So it seems like there are the options of pgroups, CEF, unpacked 8bit or unpacked 16bit
The last 2 seem to be gaining traction in implimentations as the cost of bandwidth is becoming less of an issue and the cost of processing more so.
Has anyone done an IS-11 implimentation that we could talk to/liaise with?

**Comment by kierank** [comment]
16-bit little endian please if we go down that road.

Kieran

**Comment by david3a** [comment]
Planar formats rather than interleaved is preferred by us. 

Do we need to support sub-frame assemblies such that we can send complete planar sub frames say 16 slices of an image to reduce latency, this may complicate things significantly.

Line lengths would best be paragraph aligned if possible (16 byte) for optimisation

Support Width and Stride will allow for multiple packing approaches

Simple Planar YUV422 width = 16, stride = 16, height =12 image
YYYYYYYYYYYYYYYY
-- 10 rows
YYYYYYYYYYYYYYYY

UUUUUUUU
-- 10 rows
UUUUUUUU

VVVVVVVV
-- 10 rows
VVVVVVVV

Extends to include alpha through an additional plane

AAAAAAAAAAAAAAAA
-- 10 rows
AAAAAAAAAAAAAAAA

More complex Planar YUVA 4224 width = 16, stride = 48, height =12 image
YYYYYYYYYYYYYYYYUUUUUUUUVVVVVVVVAAAAAAAAAAAAAAAA
-- 10 rows
YYYYYYYYYYYYYYYYUUUUUUUUVVVVVVVVAAAAAAAAAAAAAAAA

Second format allows for added complexity through interpretation that we would try to avoid

**Comment by jmailhot** [comment]
this has bifurcated into three topics - 
1) how to signal what "could be sent" and honor requests to switch to a specific one of them  (IS-11 plus vocabulary).  looks like we need to add to IS-11 something like "urn:x-nmos:cap:format:dataStructure".  

2) what should the enumeration of structures be (bounds of the vocabulary) - more importantly WHERE should this enumeration be located?  Let us start a list here in github and use github to modify and manage the list.  

3) what is required and what is optional.  Today, the 2110pgroups in 4:2:2 are required (see main profile), and all else is optional.  Do we need to add another required format besides pgroups?
